// firebase-tools/lib/gcp/cloudscheduler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jobFromEndpoint = exports.topicNameForEndpoint = exports.jobNameForEndpoint = exports.createOrReplaceJob = exports.updateJob = exports.getJob = exports.deleteJob = exports.createJob = void 0;
const _ = require("lodash");
const error_1 = require("../error");
const logger_1 = require("../logger");
const api_1 = require("../api");
const apiv2_1 = require("../apiv2");
const backend = require("../deploy/functions/backend");
const proto = require("./proto");
const functional_1 = require("../functional");
const VERSION = "v1";
const DEFAULT_TIME_ZONE = "America/Los_Angeles";
const apiClient = new apiv2_1.Client({ urlPrefix: api_1.cloudschedulerOrigin, apiVersion: VERSION });
function createJob(job) {
    const strippedName = job.name.substring(0, job.name.lastIndexOf("/"));
    return apiClient.post(`/${strippedName}`, Object.assign({ timeZone: DEFAULT_TIME_ZONE }, job));
}
exports.createJob = createJob;
function deleteJob(name) {
    return apiClient.delete(`/${name}`);
}
exports.deleteJob = deleteJob;
function getJob(name) {
    return apiClient.get(`/${name}`, {
        resolveOnHTTPError: true,
    });
}
exports.getJob = getJob;
function updateJob(job) {
    const fieldMasks = proto.fieldMasks(job, "pubsubTarget");
    return apiClient.patch(`/${job.name}`, Object.assign({ timeZone: DEFAULT_TIME_ZONE }, job), {
        queryParams: {
            updateMask: fieldMasks.join(","),
        },
    });
}
exports.updateJob = updateJob;
async function createOrReplaceJob(job) {
    var _a, _b;
    const jobName = job.name.split("/").pop();
    const existingJob = await getJob(job.name);
    if (existingJob.status === 404) {
        let newJob;
        try {
            newJob = await createJob(job);
        }
        catch (err) {
            if (((_b = (_a = err === null || err === void 0 ? void 0 : err.context) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.statusCode) === 404) {
                throw new error_1.FirebaseError(`Cloud resource location is not set for this project but scheduled functions require it. ` +
                    `Please see this documentation for more details: https://firebase.google.com/docs/projects/locations.`);
            }
            throw new error_1.FirebaseError(`Failed to create scheduler job ${job.name}: ${err.message}`);
        }
        logger_1.logger.debug(`created scheduler job ${jobName}`);
        return newJob;
    }
    if (!job.timeZone) {
        job.timeZone = DEFAULT_TIME_ZONE;
    }
    if (!needUpdate(existingJob.body, job)) {
        logger_1.logger.debug(`scheduler job ${jobName} is up to date, no changes required`);
        return;
    }
    const updatedJob = await updateJob(job);
    logger_1.logger.debug(`updated scheduler job ${jobName}`);
    return updatedJob;
}
exports.createOrReplaceJob = createOrReplaceJob;
function needUpdate(existingJob, newJob) {
    if (!existingJob) {
        return true;
    }
    if (!newJob) {
        return true;
    }
    if (existingJob.schedule !== newJob.schedule) {
        return true;
    }
    if (existingJob.timeZone !== newJob.timeZone) {
        return true;
    }
    if (newJob.retryConfig) {
        if (!existingJob.retryConfig) {
            return true;
        }
        if (!_.isMatch(existingJob.retryConfig, newJob.retryConfig)) {
            return true;
        }
    }
    return false;
}
function jobNameForEndpoint(endpoint, appEngineLocation) {
    const id = backend.scheduleIdForFunction(endpoint);
    return `projects/${endpoint.project}/locations/${appEngineLocation}/jobs/${id}`;
}
exports.jobNameForEndpoint = jobNameForEndpoint;
function topicNameForEndpoint(endpoint) {
    const id = backend.scheduleIdForFunction(endpoint);
    return `projects/${endpoint.project}/topics/${id}`;
}
exports.topicNameForEndpoint = topicNameForEndpoint;
function jobFromEndpoint(endpoint, appEngineLocation) {
    const job = {};
    if (endpoint.platform === "gcfv1") {
        job.name = jobNameForEndpoint(endpoint, appEngineLocation);
        job.pubsubTarget = {
            topicName: topicNameForEndpoint(endpoint),
            attributes: {
                scheduled: "true",
            },
        };
    }
    else if (endpoint.platform === "gcfv2") {
        throw new error_1.FirebaseError("Do not know how to create a scheduled GCFv2 function");
    }
    else {
        (0, functional_1.assertExhaustive)(endpoint.platform);
    }
    if (!endpoint.scheduleTrigger.schedule) {
        throw new error_1.FirebaseError("Cannot create a scheduler job without a schedule:" + JSON.stringify(endpoint));
    }
    job.schedule = endpoint.scheduleTrigger.schedule;
    job.timeZone = endpoint.scheduleTrigger.timeZone || DEFAULT_TIME_ZONE;
    if (endpoint.scheduleTrigger.retryConfig) {
        job.retryConfig = {};
        proto.copyIfPresent(job.retryConfig, endpoint.scheduleTrigger.retryConfig, "maxDoublings", "retryCount");
        proto.convertIfPresent(job.retryConfig, endpoint.scheduleTrigger.retryConfig, "maxBackoffDuration", "maxBackoffSeconds", (0, functional_1.nullsafeVisitor)(proto.durationFromSeconds));
        proto.convertIfPresent(job.retryConfig, endpoint.scheduleTrigger.retryConfig, "minBackoffDuration", "minBackoffSeconds", (0, functional_1.nullsafeVisitor)(proto.durationFromSeconds));
        proto.convertIfPresent(job.retryConfig, endpoint.scheduleTrigger.retryConfig, "maxRetryDuration", "maxRetrySeconds", (0, functional_1.nullsafeVisitor)(proto.durationFromSeconds));
        if (!Object.keys(job.retryConfig).length) {
            delete job.retryConfig;
        }
    }
    return job;
}
exports.jobFromEndpoint = jobFromEndpoint;
